File: Demos/YMusic/src/Mgr/DB/DatabaseManager.cpp
#include "stdafx.h"

#include "DatabaseManager.h"
#include "YApp.h"
#include "YMusic.h"
#include "Utility/Convert.h"
const TCHAR* CDatabaseManager::USER_DB = _T("user_%s.sl");
const TCHAR* CDatabaseManager::DEFAULT_DB = _T("default.sl");
const TCHAR* CDatabaseManager::GLOBEL_DB = _T("globel.sl");

CDatabaseManager::CDatabaseManager()
{

}

CDatabaseManager::~CDatabaseManager()
{

}

void CDatabaseManager::SetDatabasePath(const DuiLib::CDuiString& sDatabasePath)
{
	if (sDatabasePath.GetLength() == 0)
		return;
	m_sDatabasePath = sDatabasePath;
	m_sDatabasePath.Replace(_T('\\'), _T('/'));

	if (_T("/") != m_sDatabasePath.Right(1))
	{
		m_sDatabasePath += _T("/");
	}
}

DuiLib::CDuiString& CDatabaseManager::GetDatabasePath()
{
	return m_sDatabasePath;
}

bool CDatabaseManager::SetCurrentUser(spLogonUserInfoT spUserInfo)
{
	m_spUserInfo = spUserInfo;
	m_spUserDBase = spUserDBaseT(NULL);
	return true;
}
spGlobalDBaseT CDatabaseManager::GetGlobalDBase()
{
	if (m_spGlobalDBase)
		return m_spGlobalDBase;
	sLog(_T("开始创建全局数据库----数据库"));
	if (!m_spGlobalDBase && m_sDatabasePath.GetLength() > 0)
	{
		DuiLib::CDuiString sDBFileName = theFilePathMgr->GetDatabasePath() + GLOBEL_DB;
		if (!em_utility::mci::file_exist(sDBFileName))
		{

		}
		spGlobalDBaseT spGlobalDBase(new CGlobalDBase());
		if (spGlobalDBase && spGlobalDBase->OpenDB(TransToStdString(sDBFileName).c_str()))
		{
			m_spGlobalDBase = spGlobalDBase;
		}
	}
	sLog(_T("结束创建全局数据库----数据库"));
	return m_spGlobalDBase;
}
spUserDBaseT CDatabaseManager::GetUserDBase()
{
	if (m_spUserDBase)
		return m_spUserDBase;
	sLog(_T("开始创建用户列表----数据库"));
	if (!m_spUserDBase && m_spUserInfo && m_sDatabasePath.GetLength() > 0)
	{
		DuiLib::CDuiString sDBFileName = theFilePathMgr->GetDatabasePath();
		DuiLib::CDuiString sName;
		sName.Format(USER_DB, m_spUserInfo->sUserID);
		sDBFileName += sName;

		if (!em_utility::mci::file_exist(sDBFileName))
		{

		}
		spUserDBaseT spUserDBase(new CUserDBase());
		if (spUserDBase && spUserDBase->OpenDB(TransToStdString(sDBFileName).c_str()))
		{
			m_spUserDBase = spUserDBase;
		}
	}
	sLog(_T("结束创建用户列表----数据库"));
	return m_spUserDBase;
}

spUserDBaseT CDatabaseManager::GetDefaultUserDBase()
{
	if (m_spDefaultUserDBase)
		return m_spDefaultUserDBase;
	sLog(_T("开始创建默认列表----数据库"));
	if (!m_spDefaultUserDBase && m_sDatabasePath.GetLength() > 0)
	{
		DuiLib::CDuiString sDBFileName = theFilePathMgr->GetDatabasePath() + DEFAULT_DB;

		if (!em_utility::mci::file_exist(sDBFileName))
		{

		}
		spUserDBaseT spUserDBase(new CUserDBase());
		if (spUserDBase && spUserDBase->OpenDB(TransToStdString(sDBFileName).c_str()))
		{
			m_spDefaultUserDBase = spUserDBase;
		}
	}
	sLog(_T("结束创建默认列表----数据库"));
	return m_spDefaultUserDBase;
}
spUserDBaseT CDatabaseManager::GetUserDB()
{
	if (m_spUserInfo)
	{
		return GetUserDBase();
	}
	return GetDefaultUserDBase();
}

File: skin/ymusic/lua/database/DatabaseManager.lua
local CDatabaseManager = FLua.Class("CDatabaseManager")

USER_DB = "user_%s.sl"
DEFAULT_DB = "default.sl"
GLOBEL_DB = "globel.sl"
local function printf(fmt, ...)
	local v = ...
	if v then
		print(string.format(fmt, ...))
	else
		print(fmt)
	end
end
local function warnf(fmt, ...)
	local v = ...
	if v then
		warn(string.format(fmt, ...))
	else
		warn(fmt)
	end
end
local function errorf(fmt, ...)
	local v = ...
	if v then
		error(string.format(fmt, ...), 1)
	else
		error(fmt, 1)
	end
end

function CDatabaseManager:__ctor()
	self.m_spUserInfo = nil
	self.m_sDatabasePath = ""

	self.m_spGlobalDBase = nil 
	self.m_spUserDBase = nil 
	self.m_spDefaultUserDBase = nil
end

function CDatabaseManager:SetDatabasePath(sDatabasePath)
	if not sDatabasePath or string.len(sDatabasePath) == 0 then
		return
	end
	
	self.m_sDatabasePath = sDatabasePath:gsub("\\", "/")
	if self.m_sDatabasePath:sub(-1,-1) ~= "/" then
		self.m_sDatabasePath = self.m_sDatabasePath .. "/"
	end

	assert(Application.CreateDirectory(sDatabasePath))
end
function CDatabaseManager:GetDatabasePath()
	return self.m_sDatabasePath
end

function CDatabaseManager:SetCurrentUser(spUserInfo)
	self.m_spUserInfo = spUserInfo
	self.m_spUserDBase = nil
end

function CDatabaseManager:GetGlobalDBase()
	if self.m_spGlobalDBase then
		return self.m_spGlobalDBase
	end
	printf("开始创建全局数据库----数据库")
	if not self.m_spGlobalDBase and string.len(self.m_sDatabasePath) > 0 then
		local sDBFileName = self.m_sDatabasePath .. GLOBEL_DB
		local GlobalDatabaseT = require "database.GlobalDatabase"
		local spGlobalDBase = GlobalDatabaseT()
		if spGlobalDBase:OpenDB(sDBFileName) then
			self.m_spGlobalDBase = spGlobalDBase
		else
			spGlobalDBase:Release()
			spGlobalDBase = nil
		end
	end
	printf("结束创建全局数据库----数据库");
	return self.m_spGlobalDBase
end
function CDatabaseManager:GetUserDBase()
	if self.m_spUserDBase then
		return self.m_spUserDBase
	end
	printf("开始创建用户列表----数据库")
	if not self.m_spUserDBase and self.m_spUserInfo and string.len(self.m_sDatabasePath) > 0 then
		local sDBFileName = self.m_sDatabasePath;
		local sName = string.format(USER_DB, self.m_spUserInfo.sUserID)
		sDBFileName = sDBFileName .. sName

		local UserDatabaseT = require "database.UserDatabase"
		local spUserDBase = UserDatabaseT()
		if spUserDBase:OpenDB(sDBFileName) then
			self.m_spUserDBase = spUserDBase
		else
			spUserDBase:Release()
			spUserDBase = nil
		end
	end
	printf("结束创建用户列表----数据库")
	return self.m_spUserDBase
end
function CDatabaseManager:GetDefaultUserDBase()
	if self.m_spDefaultUserDBase then
		return self.m_spDefaultUserDBase
	end
	printf("开始创建默认列表----数据库")
	if not self.m_spDefaultUserDBase and string.len(self.m_sDatabasePath) > 0 then
		local sDBFileName = self.m_sDatabasePath .. DEFAULT_DB
		
		local UserDatabaseT = require "database.UserDatabase"
		local spUserDBase = UserDatabaseT()
		if spUserDBase:OpenDB(sDBFileName) then
			self.m_spDefaultUserDBase = spUserDBase
		else
			spUserDBase:Release()
			spUserDBase = nil
		end
	end
	printf("结束创建默认列表----数据库")
	return self.m_spDefaultUserDBase
end

function CDatabaseManager:GetUserDB()
	if self.m_spUserInfo then
		return self:GetUserDBase()
	end
	return self:GetDefaultUserDBase()
end

return CDatabaseManager